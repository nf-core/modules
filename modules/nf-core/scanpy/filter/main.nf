process SCANPY_FILTER {
    tag "${meta.id}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/16/168ecbbe27ccef766741ccbf937b0d2675be2e19b0565035e0719f1e9ea5ee95/data'
        : 'community.wave.seqera.io/library/python_pyyaml_scanpy:b5509a698e9aae25'}"

    input:
    tuple val(meta), path(h5ad)
    val min_genes
    val min_cells
    val min_counts_gene
    val min_counts_cell
    val max_mito_percentage
    val symbol_col

    output:
    tuple val(meta), path("*.h5ad"), emit: h5ad
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    prefix     = task.ext.prefix ?: "${meta.id}_filtered"
    if ("${prefix}.h5ad" == "${h5ad}") {
        error("Input and output names are the same, use \"task.ext.prefix\" to disambiguate!")
    }
    template('filter.py')

    stub:
    prefix = task.ext.prefix ?: "${meta.id}_filtered"
    if ("${prefix}.h5ad" == "${h5ad}") {
        error("Input and output names are the same, use \"task.ext.prefix\" to disambiguate!")
    }
    """
    # These are needed to prevent errors during import of scanpy
    # when using singularity/apptainer
    export MPLCONFIGDIR=./tmp/mpl
    export NUMBA_CACHE_DIR=./tmp/numba

    touch ${prefix}.h5ad

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        scanpy: \$(python3 -c 'import scanpy; print(scanpy.__version__)')
    END_VERSIONS
    """
}
