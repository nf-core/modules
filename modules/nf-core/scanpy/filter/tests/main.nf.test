nextflow_process {

    name "Test Process SCANPY_FILTER"
    script "../main.nf"
    process "SCANPY_FILTER"

    tag "modules"
    tag "modules_nfcore"
    tag "scanpy"
    tag "scanpy/filter"

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = Channel.of([
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/scrnaseq/h5ad/SRR28679756_filtered_matrix.h5ad', checkIfExists: true)
                    ]
                )
                input[1] = 20
                input[2] = 20
                input[3] = 50
                input[4] = 50
                input[5] = 50
                input[6] = 'index'
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml,
                ).match() }
            )
        }

    }

    test("Should run with a different symbol column") {

        when {
            process {
                """
                input[0] = Channel.of([
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/scrnaseq/h5ad/SRR28679756_filtered_matrix.h5ad', checkIfExists: true)
                    ]
                )
                input[1] = 20
                input[2] = 20
                input[3] = 50
                input[4] = 50
                input[5] = 50
                input[6] = 'symbols'
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml,
                ).match() }
            )
        }

    }

    test("Should fail with a non-existent symbol column") {

        when {
            process {
                """
                input[0] = Channel.of([
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/scrnaseq/h5ad/SRR28679756_filtered_matrix.h5ad', checkIfExists: true)
                    ]
                )
                input[1] = 20
                input[2] = 20
                input[3] = 50
                input[4] = 50
                input[5] = 50
                input[6] = 'non_existent_column'
                """
            }
        }

        then {
            assertAll(
            { assert process.failed }
            )
        }

    }

    test("Should run without failures - stub") {

        options '-stub'

        when {
            process {
                """
                input[0] = Channel.of([
                        [ id: 'test' ],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/scrnaseq/h5ad/SRR28679756_filtered_matrix.h5ad', checkIfExists: true)
                    ]
                )
                input[1] = 20
                input[2] = 20
                input[3] = 50
                input[4] = 50
                input[5] = 50
                input[6] = 'index'
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml,
                ).match() }
            )
        }

    }

}
