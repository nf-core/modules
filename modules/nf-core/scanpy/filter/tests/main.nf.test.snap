{
    "Should run with a different symbol column": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.h5ad:md5,5e4f50b4961a1d67a1091056868f34a0"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a2e4b6b01deba3f07dce478c56acb14d"
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.h5ad:md5,5e4f50b4961a1d67a1091056868f34a0"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a2e4b6b01deba3f07dce478c56acb14d"
                ]
            },
            {
                "SCANPY_FILTER": {
                    "python": "3.12.11",
                    "scanpy": "1.11.2"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-04T09:48:00.155846767"
    },
    "Should run without failures - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,8bcc268104b7c43e2be4e678081d0596"
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8bcc268104b7c43e2be4e678081d0596"
                ]
            },
            {
                "SCANPY_FILTER": {
                    "python": "3.12.11",
                    "scanpy": "1.11.2"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-25T10:28:53.871079892"
    },
    "Should run without failures": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.h5ad:md5,5e4f50b4961a1d67a1091056868f34a0"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a2e4b6b01deba3f07dce478c56acb14d"
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_filtered.h5ad:md5,5e4f50b4961a1d67a1091056868f34a0"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a2e4b6b01deba3f07dce478c56acb14d"
                ]
            },
            {
                "SCANPY_FILTER": {
                    "python": "3.12.11",
                    "scanpy": "1.11.2"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-04T09:47:51.215629629"
    }
}